hpack/table.py

```diff
@@ -0,0 +1,215 @@
+            return HeaderTable.STATIC_TABLE[index]
+        index -= len(HeaderTable.STATIC_TABLE)
+        if index < len(self):
+            return deque.__getitem__(self, index)
+        return None # TODO throw HPACKException here
+
+    def __setitem__(self, index, value):
+        """
+        We don't support direct index setting
+        """
+        raise TypeError("'HeaderTable' object does not support item assignment, use add")
+
+    def __repr__(self):
+        rv = "HeaderTable("
+        rv += str(self._maxsize)
+        rv += ", "+str(self.resized)
+        rv += ", ["
+        itm = [str(entry) for entry in self]
+        itm = ", ".join(itm)
+        return rv+itm+"])"
+
+    def add(self, name, value):
+        """
+        Adds a new entry to the table
+
+        We reduce the table size if the entry will make the
+        table size greater than maxsize.
+        """
+        # We just clear the table if the entry is to big
+        if self._entry_size(name, value) > self._maxsize:
+          self.clear()
+        # Add new entry if the table actually has a size
+        elif self._maxsize > 0:
+          self.appendleft((name,value))
+          self._shrink()
+
+    def search(self, name, value):
+        """
+        Searchs the table for the entry spefied by name and
+        value
+
+        Returns one of the following:
+            None                 no match at all
+            (index, name, None)  partial match (on name)
+            (index, name, value) perfect match
+        """
+        offset = len(HeaderTable.STATIC_TABLE)
+        partial = None
+        for (i, (n, v)) in enumerate(HeaderTable.STATIC_TABLE):
+            if n == name:
+                if v == value:
+                    return (i + 1, n, v)
+                elif partial is None:
+                    partial = (i + 1, n, None)
+        for (i, (n, v)) in enumerate(self):
+            if n == name:
+                if v == value:
+                    return (i + offset + 1, n, v)
+                elif partial is None:
+                    partial = (i + offset + 1, n, None)
+        return partial
+
+    @property
+    def maxsize(self):
+        return self._maxsize
+
+    @maxsize.setter
+    def maxsize(self, newmax):
+        newmax = int(newmax)
+        oldmax = self._maxsize
+        self._maxsize = newmax
+        self.resized = (newmax != oldmax)
+        if newmax <= 0:
+          self.clear()
+        elif oldmax > newmax:
+          self._shrink()
+
+    def _entry_size(self, name, value):
+        """
+        Calculates the size of a single entry
+
+        This size is mostly irrelevant to us and defined
+        specifically to accomidate memory management for
+        lower level implementions. The 32 extra bytes are
+        considered the "maximum" overhead that would be
+        required to represent each entry in the table.
+
+        See RFC7541 Section 4.1
+        """
+        return 32 + len(name) + len(value)
```

Comment:
This is not a method. Things should only be a method if the keyword `self` appears in them. If they don't, they should be functions.